text/0047-storage.md

```diff
@@ -0,0 +1,217 @@
+/// which determine how to start and shutdown the storage
+pub trait Backend {
+    /// start method should impl how to start and initialize the corrsponding database
+    /// It takes config_path which define the database options, and returns Result<Self, Box<dyn Error>>
+    async fn start(config_path: String) -> Result<Self, Box<dyn Error>> where Self: Sized;
+    /// shutdown method should impl how to shutdown the corrsponding database
+    /// It takes the ownership of self, and returns ()
+    async fn shutdown(self) -> Result<(), Box<dyn Error>>;
+}
+#[cfg(feature = "rocks_db")]
+#[async_trait]
+impl Backend for Storage {
+    /// It starts RocksDB instance and then initialize the required column familes
+    async fn start(config_path: String) -> Result<Self, Box<dyn Error>> {
+        let config_as_string = fs::read_to_string(config_path)?;
+        let config: config::Config = toml::from_str(&config_as_string)?;
+        let db = rocksdb::RocksdbBackend::new(config)?;
+        Ok(Storage { inner: db })
+    }
+    /// It shutdown RocksDB instance,
+    /// Note: the shutdown is done through flush method and then droping the storage object
+    async fn shutdown(self) -> Result<(), Box<dyn Error>> {
+        self.inner.flush()?
+    }
+}
+```
+
+- Operations macros which should be imported in the associated object mod in other crates
+    - impl_transaction_ops!(BundleTransaction);
+    - impl_milestone_ops!(Milestone);
+    - impl_tangle_ops!(Tangle);
+    - impl_.._ops!(..);
+
+```rust
+// Using macros is how bee_storage expose database operations
+#[macro_export]
+macro_rules! impl_transaction_ops {
+    ($object:ty) => {
+        use bee_storage::storage::rocksdb::TRANSACTION_CF_HASH_TO_TRANSACTION;
+        use bee_transaction::bundled::{BundledTransaction,BundledTransactionField};
+        use bee_ternary::{T1B1Buf, T5B1Buf, TritBuf, Trits, T5B1};
+        use bytemuck::cast_slice;
+        use bee_crypto::ternary::Hash;
+        use bee_storage::access::OpError;
+        #[cfg(feature = "rocks_db")]
+        impl $object {
+            async fn insert(&self, hash: Hash, storage: &Storage) -> Result<(), OpError> {
+                // get column family handle to hash_to_tx table in order presist the transaction;
+                let hash_to_tx = storage.inner.cf_handle(TRANSACTION_CF_HASH_TO_TRANSACTION).unwrap();
+                let mut tx_trit_buf = TritBuf::<T1B1Buf>::zeros(Self::trit_len());
+                self.into_trits_allocated(tx_trit_buf.as_slice_mut());
+                let hash_buf = hash.to_inner().encode::<T5B1Buf>();
+                storage.inner.put_cf(
+                    &hash_to_tx,
+                    cast_slice(hash_buf.as_i8_slice()),
+                    cast_slice(tx_trit_buf.encode::<T5B1Buf>().as_i8_slice()),
+                )?;
+                Ok(())
+            }
+            async fn find_by_hash(hash: Hash, storage: &Storage) -> Result<Option<Self>, OpError> {
```

Comment:
`Hash` should be passed by reference to avoid unnecessary cloning (I think `Hash` implements `Deref<Trits>` now, so you can just call `.encode()` on it directly.

Reply:
Perfect, yeah I will force reference everywhere.